### PR TITLE
WIP: `json_encode`, and `vim_snprintf`, loses floating point precision when using `%g`

### PR DESCRIPTION
**Pull request by S0AndS0**
:warning: This breaks `echo` and messes with other Vim Ex mode commands, ex;

```vim
echo json_encode({'value':0.5670403838157654})
```
> Results could be;
>
> - `{"value":5.068898e-310}`
> - `{"value":1.237282e-313}`
> - `{"value":0.0}`

```vim
:echo json_decode('{"value":0.5670403838157654}')
```
> Results, regardless of float size, always be;
>
> - `{'value': 0.0}`

... Side note; `printf` within Vim Ex mode also has inconsistent behaviors.

## Questions

- Is this worth pursuing?  And if so;
  - What causes the inconsistencies?
  - Where else do I need to apply updates?
  - Is there a convenient way to avoid/expand `e` notation?
  - Any pointers on adding some unit tests?

## Story time

I'm writing a plugin, aimed at Vim8 compatibility, that makes calls to an API that I do not control outputs of.  And the API returns JSON that contains large floats, example;

```json
{
  "embedding": [
    0.5670403838157654,
    0.009260174818336964,
    0.23178744316101074,
    -0.2916173040866852,
    -0.8924556970596313,
    0.8785552978515625,
    -0.34576427936553955,
    0.5742510557174683,
    -0.04222835972905159,
    -0.137906014919281
  ]
}
```

... As of Vim version `9.1.1-785` packaged by Arch, I use Arch (BTW™), these values are truncated to about five digits past the decimal point, eg.

```vim
:echo json_decode('{"embedding":[0.5670403838157654]}')
```
> Result: `{'embedding': [0.56704]}`

Oddly it looks like this might be a `echo` bug, because wrapping with `printf` and using a large limit mostly preserves precision of input floats... _mostly_

```vim
:echo printf('%.25g', json_decode('{"embedding":[0.5670403838157654]}').embedding[0])
```
> Result: `0.5670403838157653808593750`
>
> Notice, we lost the last `4` and picked-up a bunch of extra `3808593750` :shrug:

### CODE REVIEW

**Comment by tonymec** [comment]
I can't find it back in the Vim help, but IIRC `:echo` has always displayed Floats with some standardized default precision of 6 significant digits or so regardless of what the hardware could handle. With `printf()`, OTOH, it is possible to specify the precision, including a better precision than what the hardware affords, in which case "anything" will be displayed beyond the hardware limit. For instance if you try to use `printf()` to display `1.0 / 3.0` with tremendous precision, you'll notice that at some point it ceases to be a string of repetitions of the digit 3.

**Comment by chrisbra** [comment]
It currently is at `:h floating-point-precision`

**Comment by tonymec** [comment]
> It currently is at `:h floating-point-precision`

Ah, thanks. Yes, this is what I meant.

**Comment by S0AndS0** [comment]
I agree that `echo` be a non-issue to worry about, and the main thing I'm currently concerned with is preserving precision whenever possible; especially in regards to JSON encode/decode operations.

Double-checking myself on Vim version `9.1.1-785`, after learning that `echo` for sure cannot be trusted, it looks like the decoding is fine, ex.

```vim
let json_stringy = '{ "value": "0.009260174818336964" }'
let json_floater = '{ "value": 0.009260174818336964 }'

let parsed_stringy = printf('%s', json_decode(json_stringy).value)
let parsed_floater = printf('%.18g', json_decode(json_floater).value)

echo parsed_stringy == parsed_floater
"> 1
echo len(parsed_stringy)
"> 20
echo len(parsed_floater)
"> 20
```

...  But things get funky when re-encoding;

```vim
echo json_encode(json_decode(json_stringy))
"> {"value":"0.009260174818336964"}

echo json_encode(json_decode(json_floater))
"> {"value":0.00926}
```

So I may need to revert previously proposed changes and instead fixate on what `json_encode` is doing to floats.

**Comment by zeertzjq** [comment]
Please do not change version.c

**Comment by S0AndS0** [comment]
I think I got most of the bits for allowing `long double` (AKA _long floats_) usage!

Only bit that is being a stinker are print/echo-ing dictionaries as a whole, eg.

```vim
echo {'floater': 0.419}
"> `{'floater': 0.00000000000000000000}`
echo {'floater': 0.419}.floater
"> `6.95323265041282803632e-310`
```

...  But if feels close to doable, closer at least, and it seems `printf` is now properly passing the long variants of floats;

```vim
echo printf('%.18Lg', {'value':0.00926017481833696365}.value)
"> `0.009260174818336964`
```

**TLDR** I'd much appreciate a pointer to where I'm messing thing up when a dictionary''

**Comment by S0AndS0** [comment]
> Please do not change version.c in a PR

Uh-oh :facepalm: ...  I'll try to get that reverted _soon_

**Comment by chrisbra** [comment]
this has stalled, so closing
